src/lib/transaction.rs

```diff
@@ -0,0 +1,586 @@
+                (&mut buf[n..]).write(&nul.to_bytes()).and_then(|num| {
+                    n += num;
+                    Ok(num)
+                })
+            })
+            .collect::<io::Result<Vec<usize>>>()?;
+
+        let crossover_present = self.crossover.is_some() as u8;
+        n += (&mut buf[n..]).write(&crossover_present.to_le_bytes())?;
+        if crossover_present != 0 {
+            n += self.crossover.unwrap().read(&mut buf[n..])?;
+        }
+        n += (&mut buf[n..]).write(&(self.notes.len() as u64).to_le_bytes())?;
+
+        self.notes
+            .iter_mut()
+            .map(|note| {
+                note.read(&mut buf[n..]).and_then(|num| {
+                    n += num;
+                    Ok(num)
+                })
+            })
+            .collect::<io::Result<Vec<usize>>>()?;
+
+        n += self.fee.read(&mut buf[n..])?;
+
+        let proof_bytes = self.spending_proof.to_bytes();
+        n += (&mut buf[n..]).write(&proof_bytes)?;
+
+        n += (&mut buf[n..])
+            .write(&(self.call_data.len() as u64).to_le_bytes())?;
+        n += (&mut buf[n..]).write(&self.call_data)?;
+
+        Ok(n)
+    }
+}
+
+impl Write for TransactionPayload {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut n = 0;
+
+        let mut one_scalar = [0u8; 32];
+        let mut one_u64 = [0u8; 8];
+        let mut one_note = [0u8; 233];
+        let mut one_byte = [0u8; 1];
+
+        n += (&buf[n..]).read(&mut one_scalar)?;
+        self.anchor = Option::from(BlsScalar::from_bytes(&one_scalar)).ok_or(
+            io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Could not deserialize anchor",
+            ),
+        )?;
+
+        n += (&buf[n..]).read(&mut one_u64)?;
+        let nul_size = u64::from_le_bytes(one_u64) as usize;
+        self.nullifiers = Vec::<BlsScalar>::with_capacity(nul_size);
+        (0..nul_size)
+            .into_iter()
+            .map(|_| {
+                (&buf[n..]).read(&mut one_scalar).and_then(|num| {
+                    n += num;
+                    self.nullifiers.push(
+                        Option::from(BlsScalar::from_bytes(&one_scalar))
+                            .ok_or(io::Error::new(
+                                io::ErrorKind::InvalidData,
+                                "Could not deserialize nullifier",
+                            ))?,
+                    );
+
+                    Ok(n)
+                })
+            })
+            .collect::<Result<Vec<usize>, io::Error>>()?;
+
+        n += (&buf[n..]).read(&mut one_byte)?;
+        if u8::from_le_bytes(one_byte) != 0 {
+            let mut crossover = Crossover::default();
+            n += crossover.write(&buf[n..])?;
+            self.crossover = Some(crossover);
+        }
+
+        n += (&buf[n..]).read(&mut one_u64)?;
+        let notes_size = u64::from_le_bytes(one_u64) as usize;
+        self.notes = vec![Note::default(); notes_size];
```

Comment:
Here you're doing some calculation creating default notes that I believe are not necessary since they're going to be replaced by the actual ones, right?

I will open an issue on phoenix-core to implement the to / from bytes, that should simplify the code here and avoid the computation for the default's Note.